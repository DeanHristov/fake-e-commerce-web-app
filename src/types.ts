export enum PRODUCT_CATEGORIES {
  ELECTRONIC = 'ELECTRONIC',
}

export type TCurrency = 'USD' | 'BGN' | 'EUR';

export interface IComment {
  productId: string;
  _id: string;
  name: string;
  email: string;
  body: string;
  date: string;
}

// Object schema can be seen here: https://dummyjson.com/docs/products
export interface IProduct {
  id: number;
  title: string;
  description: string;
  price: number;
  discountPercentage: number;
  rating: number;
  stock: number;
  brand: string;
  category: string;
  thumbnail: string;
  images?: string[];
  comments?: IComment[];
}

export interface IUser {
  name: string;
  email: string;
  password: string;
  isAdmin: boolean;
}

export interface ICartProduct {
  id: number;
  price: number;
  quantity: number;
  discountPercentage?: number;
  // TODO Fix next line!
  discountedPrice?: number;
  title: string;
  thumbnail: string;
  total: number;
  inStock: number;
}

export interface ICart {
  id: number;
  userId: number;
  total: number;
  discountedTotal: number;
  totalProducts: number;
  totalQuantity: number;
  currency?: TCurrency;
  products: Record<number, ICartProduct>; // Map<number, ICartProduct>;
}

//@see:https://nextjs.org/docs/app/api-reference/functions/generate-metadata#generatemetadata-function
export interface IMetadataPageTitle {
  title: string;
}

export interface IProductsResponse {
  products: IProduct[];
  total: number;
  skip: number;
  limit: number;
}

export interface ICartsResponse {
  carts: ICart[];
  total: number;
  skip: number;
  limit: number;
}
