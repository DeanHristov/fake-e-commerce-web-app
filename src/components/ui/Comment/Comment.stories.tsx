import type { Meta, StoryObj } from '@storybook/react';

import Comment, { ICommentProps } from './Comment';

//TODO Fix next/image issue in the storybook!
const meta: Meta<ICommentProps> = {
  title: 'UI/Comment',
  component: Comment,
  parameters: {
    layout: 'centered',
  },
};

type Story = StoryObj<ICommentProps>;

export const Primary: Story = {
  args: {
    name: 'John Doue',
    date: '20 April 2022, at 14:88 PM',
    body: 'It is a great product!',
  },
};

export default meta;
