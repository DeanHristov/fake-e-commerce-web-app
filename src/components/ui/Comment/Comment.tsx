import Image from 'next/image';
import { FC } from 'react';
import { IComment } from '../../../types';

export interface ICommentProps extends Omit<IComment, 'productId' | 'email'> {}

const Comment: FC<ICommentProps> = ({ _id, name, body, date }) => {
  return (
    <div className="relative grid grid-cols-1 gap-4 p-4 mb-8 border rounded-lg bg-white">
      <div className="relative flex gap-4">
        <Image
          src="/images/avatar.jpeg"
          width={64}
          height={20}
          className="relative rounded-lg -top-8 -mb-4 bg-white border"
          alt=""
        />
        <div className="flex flex-col w-full">
          <div className="flex flex-row justify-between">
            <p className="relative text-xl whitespace-nowrap truncate overflow-hidden">
              {name}
            </p>
            <a className="text-gray-500 text-xl" href="#">
              <i className="fa-solid fa-trash"></i>
            </a>
          </div>
          <p className="text-gray-400 text-sm">{date}</p>
        </div>
      </div>
      <p className="-mt-4 text-gray-500">{body}</p>
    </div>
  );
};

export default Comment;
