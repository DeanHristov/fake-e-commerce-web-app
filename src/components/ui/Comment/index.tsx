import Comment, { ICommentProps } from './Comment';

export type { ICommentProps };

export default Comment;
