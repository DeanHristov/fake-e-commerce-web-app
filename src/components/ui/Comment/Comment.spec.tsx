import { render } from '@testing-library/react';
import { IComment } from '../../../types';
import Comment from './Comment';

const mockComment: IComment = {
  productId: '29a34b48-8343-4e3d-8c21-3548ba249936',
  _id: '29a34b48-8343-4e3d-8c21',
  name: 'John Doue',
  date: '20 April 2022, at 14:88 PM',
  email: 'john-doue@example.com',
  body: 'It is a great product!',
};

describe('UI/Component <Comment {...} />', () => {
  it('', () => {
    const { container } = render(<Comment {...mockComment} />);

    expect(container).toMatchSnapshot();
  });
});
