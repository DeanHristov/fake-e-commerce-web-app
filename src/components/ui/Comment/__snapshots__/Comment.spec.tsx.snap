// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI/Component <Comment {...} /> Should render a default state of the component 1`] = `
<div>
  <div
    class="relative grid grid-cols-1 gap-4 p-4 mb-8 border rounded-lg bg-white"
  >
    <div
      class="relative flex gap-4"
    >
      <img
        alt=""
        class="relative rounded-lg -top-8 -mb-4 bg-white border"
        data-nimg="1"
        decoding="async"
        height="20"
        loading="lazy"
        src="/_next/image?url=%2Fimages%2Favatar.jpeg&w=128&q=75"
        srcset="/_next/image?url=%2Fimages%2Favatar.jpeg&w=64&q=75 1x, /_next/image?url=%2Fimages%2Favatar.jpeg&w=128&q=75 2x"
        style="color: transparent;"
        width="64"
      />
      <div
        class="flex flex-col w-full"
      >
        <div
          class="flex flex-row justify-between"
        >
          <p
            class="relative text-xl whitespace-nowrap truncate overflow-hidden"
          >
            John Doue
          </p>
          <a
            class="text-gray-500 text-xl"
            href="#"
          >
            <i
              class="fa-solid fa-trash"
            />
          </a>
        </div>
        <p
          class="text-gray-400 text-sm"
        >
          20 April 2022, at 14:88 PM
        </p>
      </div>
    </div>
    <p
      class="-mt-4 text-gray-500"
    >
      It is a great product!
    </p>
  </div>
</div>
`;
